plugins/org.python.pydev.ast/src/org/python/pydev/ast/codecompletion/PyStringCodeCompletion.java

```diff
@@ -149,17 +159,55 @@
     /**
      * Needed interface for adding the completions on a request
      * @throws MisconfigurationException
+     * @throws PythonNatureWithoutProjectException
+     * @throws IOException
      */
     @Override
     public TokensOrProposalsList getCodeCompletionProposals(CompletionRequest request) throws CoreException,
-            BadLocationException, MisconfigurationException {
+            BadLocationException, MisconfigurationException, IOException, PythonNatureWithoutProjectException {
+
         List<ICompletionProposalHandle> completionProposals = new ArrayList<>();
         request.showTemplates = false; //don't show templates in strings
         fillWithEpydocFields(request, completionProposals);
 
         TokensOrProposalsList ret = new TokensOrProposalsList();
         if (completionProposals.size() == 0) {
             //if the size is not 0, it means that this is a place for the '@' stuff, and not for the 'default' context for a string.
+            IDocument doc = request.doc;
+            ITypedRegion partition = ((FastPartitioner) PyPartitionScanner.checkPartitionScanner(doc))
+                    .getPartition(request.documentOffset);
+            String partitionType = partition.getType();
+
+            if (IPythonPartitions.F_STRING_PARTITIONS.contains(partitionType)) {
+                // Now we are going to check whether where we are in the given completion offset
+                int requestOffset = request.documentOffset;
+                int fStringLine = doc.getLineOfOffset(partition.getOffset());
+                int fStringLineOffset = doc.getLineOffset(fStringLine);
+
+                String str = doc.get(fStringLineOffset,
+                        partition.getOffset() + partition.getLength() - fStringLineOffset);
+
+                FStringsAST ast = null;
+                try {
+                    ast = FStringsGrammarFactory.createGrammar(str).f_string();
+                } catch (Throwable e) {
+                    // Just ignore any errors for this.
+                }
+
+                if (ast != null && ast.hasChildren()) {
+                    for (SimpleNode fStringExpression : ast.getFStringExpressions()) {
+                        int fStringExpressionOffset = doc.getLineOffset(fStringLine + fStringExpression.beginLine - 1)
```

Comment:
This logic seems a bit odd to be... If you want to get the start/end offset of a node, you can do something as:

```
int nodeStartOoffset = TextSelectionUtils.getAbsoluteCursorOffset(doc, node.beginLine - 1, node.beginColumn - 1);
int nodeEndOffset = TextSelectionUtils.getAbsoluteCursorOffset(doc, node.endLine - 1, node.endColumn);
```

Reply:
When creating the grammar using `String str`, all the offsets is based on the string length/offset, not on the doc content. So I needed to do all this handle, because the ast always starts at line 1, offset 1 (1-based)

Reply:
Then it'd be just adding that line diff to the node...

Something as:

```
int nodeStartOoffset = TextSelectionUtils.getAbsoluteCursorOffset(doc, fStringLine+node.beginLine - 1, fStringLine+node.beginColumn - 1);
int nodeEndOffset = TextSelectionUtils.getAbsoluteCursorOffset(doc, fStringLine+node.endLine - 1, fStringLine+node.endColumn);
```

Reply:
Actually, you also need the column offset to be added.

Reply:
p.s.: The column needs to be added only if the node line == ast start line.